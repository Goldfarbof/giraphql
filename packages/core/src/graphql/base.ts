import { GraphQLType } from 'graphql';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default abstract class BaseType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string | keyof Types['Input'] | keyof Types['Output'],
  Shape,
  InputShape = Shape,
  MatchShape = Shape
> {
  typename: Name;

  shape?: Shape;

  inputShape?: InputShape;

  matchShape?: MatchShape;

  abstract kind: 'Object' | 'Union' | 'Interface' | 'Enum' | 'Scalar' | 'InputObject';

  constructor(name: Name, shape?: Shape) {
    this.typename = name;
    this.shape = shape;
  }

  abstract buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLType;
}
