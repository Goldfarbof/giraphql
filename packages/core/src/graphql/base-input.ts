import BaseType from './base';

export default abstract class BaseInputType<Shape = unknown, InputShape = Shape> extends BaseType<
  Shape
> {
  shape?: Shape;

  inputShape?: InputShape;

  abstract kind: 'Enum' | 'Scalar' | 'InputObject';
}
