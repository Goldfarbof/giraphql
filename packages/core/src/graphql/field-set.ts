import { FieldsShape } from '../types';

export default class FieldSet<Types extends GiraphQLSchemaTypes.TypeInfo, Shape = unknown> {
  kind: 'FieldSet' = 'FieldSet';

  forType: string;

  shape: FieldsShape<any, any>;

  constructor(type: string, shape: FieldsShape<Types, Shape>) {
    this.forType = type;
    this.shape = shape;
  }
}
