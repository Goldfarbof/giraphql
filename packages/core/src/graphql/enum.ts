import { GraphQLEnumType, GraphQLEnumValueConfigMap, GraphQLEnumValueConfig } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { EnumValues, ShapeFromEnumValues } from '../types';

export default class EnumType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string,
  Values extends EnumValues = EnumValues,
  Shape extends string = ShapeFromEnumValues<Values>
> extends BaseType<Types, Name, Shape> {
  values: GraphQLEnumValueConfigMap;

  description?: string;

  kind: 'Enum' = 'Enum';

  options: GiraphQLSchemaTypes.EnumTypeOptions<EnumValues>;

  constructor(name: Name, options: GiraphQLSchemaTypes.EnumTypeOptions<Values>) {
    super(name);

    this.options = options;

    this.values = Array.isArray(options.values)
      ? fromEntries(options.values.map(key => [key, {}]))
      : fromEntries(
          Object.entries(options.values)
            .map(([key, value]) => {
              if (value && typeof value === 'object') {
                return [key, value];
              }

              if (typeof value === 'string') {
                return [value, {}];
              }

              return null;
            })
            .filter(Boolean) as [string, GraphQLEnumValueConfig][],
        );

    this.description = options.description;
  }

  buildType() {
    return new GraphQLEnumType({
      description: this.description,
      name: this.typename,
      values: this.values,
    });
  }
}
