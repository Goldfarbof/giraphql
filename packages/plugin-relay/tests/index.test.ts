import { gql } from 'apollo-server';
import { execute, printSchema } from 'graphql';
import schemaWithGlobalConnectionFields from './examples/global-connection-fields/schema';
import schema from './examples/relay/schema';

describe('relay example schema', () => {
  it('generates expected schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('generates expected schema with globalConnectionFields', () => {
    expect(printSchema(schemaWithGlobalConnectionFields)).toMatchSnapshot();
  });

  describe('queries', () => {
    it('query offset based connections', async () => {
      const query = gql`
        {
          numbers(first: 3) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfterAtEnd: numbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjoxOTc=") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfter: numbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastAtBegining: numbers(before: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastBefore: numbers(last: 3, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          beforeAndAfter: numbers(
            before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA=="
            after: "T2Zmc2V0Q29ubmVjdGlvbjo3"
          ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAndLast: numbers(first: 5, last: 2, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          sharedConnection {
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    it('query array based connections', async () => {
      const query = gql`
        {
          batchNumbers(first: 3) {
            connectionField
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              edgeField
              cursor
              node {
                number
              }
            }
          }
          firstAfterAtEnd: batchNumbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjoxOTc=") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfter: batchNumbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastAtBegining: batchNumbers(before: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastBefore: batchNumbers(last: 3, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          beforeAndAfter: batchNumbers(
            before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA=="
            after: "T2Zmc2V0Q29ubmVjdGlvbjo3"
          ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAndLast: batchNumbers(first: 5, last: 2, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    it('query node and nodes on query type', async () => {
      const query = gql`
        {
          extraNode {
            __typename
            ... on Number {
              number
            }
          }
          moreNodes {
            __typename
            ... on BatchNumber {
              batchNumber: number
            }
            ... on Number {
              number
            }
          }
          nodes(ids: ["TnVtYmVyOjQ=", "TnVtYmVyOjI=", "TnVtYmVyOjI=", "QmF0Y2hOdW1iZXI6MA=="]) {
            ... on BatchNumber {
              batchNumber: number
            }
            ... on Number {
              number
            }
          }

          batchNode: node(id: "QmF0Y2hOdW1iZXI6MA==") {
            ... on BatchNumber {
              number
            }
          }
          node(id: "TnVtYmVyOjI=") {
            id
            ... on Number {
              number
            }
          }

          node2: node(id: "TnVtYmVyOjI=") {
            ... on Number {
              number
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    it('globalID inputs', async () => {
      const query = gql`
        {
          inputGlobalID(
            id: "YWJjOjEyMw=="
            normalId: 123
            inputObj: {
              id: "YWJjOjEyMw=="
              idList: ["YWJjOjEyMw=="]
              circular: { id: "YWJjOjEyMw==", idList: ["YWJjOjEyMw=="] }
            }
          )
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(JSON.parse(result.data!.inputGlobalID)).toMatchInlineSnapshot(`
        Object {
          "id": Object {
            "id": "123",
            "typename": "abc",
          },
          "inputObj": Object {
            "circular": Object {
              "id": Object {
                "id": "123",
                "typename": "abc",
              },
              "idList": Array [
                Object {
                  "id": "123",
                  "typename": "abc",
                },
              ],
            },
            "id": Object {
              "id": "123",
              "typename": "abc",
            },
            "idList": Array [
              Object {
                "id": "123",
                "typename": "abc",
              },
            ],
          },
          "normal": "123",
        }
      `);
    });
  });

  describe('mutations', () => {
    it('returns correct clientMutationId', async () => {
      const query = gql`
        mutation {
          a: exampleMutation(input: { clientMutationId: "a", id: "123" }) {
            clientMutationId
            itWorked
          }
          b: exampleMutation(input: { clientMutationId: "b", id: "345" }) {
            clientMutationId
            itWorked
          }
          exampleWithDescriptions(customInput: { clientMutationId: "c", id: "123" }) {
            clientMutationId
            itWorked
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "a": Object {
              "clientMutationId": "a",
              "itWorked": true,
            },
            "b": Object {
              "clientMutationId": "b",
              "itWorked": false,
            },
            "exampleWithDescriptions": Object {
              "clientMutationId": "c",
              "itWorked": true,
            },
          },
        }
      `);
    });
  });
});
