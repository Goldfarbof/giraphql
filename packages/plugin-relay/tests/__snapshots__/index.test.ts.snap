// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relay example schema generates expected schema 1`] = `
"type Answer {
  count: Int!
  id: ID!
  value: String!
}

type BatchNumber implements Node {
  id: ID!
  number: Int!
}

\\"\\"\\"input type\\"\\"\\"
input CustomInputName {
  clientMutationId: ID!
  id: ID!
}

\\"\\"\\"output type\\"\\"\\"
type CustomOutputName {
  clientMutationId: ID!
  itWorked: Boolean!
}

input ExampleMutationInput {
  clientMutationId: ID!
  id: ID!
}

type ExampleMutationPayload {
  clientMutationId: ID!
  itWorked: Boolean!
}

input GlobalIDInput {
  circular: GlobalIDInput
  circularWithoutGlobalIds: NoGlobalIDInput
  id: ID!
  idList: [ID!]!
}

type Mutation {
  answerPoll(answer: Int!, id: ID!): Poll!
  createPoll(answers: [String!]!, question: String!): Poll!
  exampleMutation(input: ExampleMutationInput!): ExampleMutationPayload!

  \\"\\"\\"mutation field\\"\\"\\"
  exampleWithDescriptions(customInput: CustomInputName!): CustomOutputName!
}

input NoGlobalIDInput {
  circular: NoGlobalIDInput
  id: ID
}

interface Node {
  id: ID!
}

type Number implements Node {
  id: ID!
  number: Int!
}

type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

type Poll implements Node {
  answers: [Answer!]!
  answersConnection(after: ID, before: ID, first: Int, last: Int): PollAnswersConnection!
  answersUsingOffset(after: ID, before: ID, first: Int, last: Int): PollAnswersUsingOffsetConnection!
  answersWithoutHelpers(after: ID, before: ID, first: Int, last: Int): PollAnswersCon!
  id: ID!
  question: String!
  updatedAt: String!
}

type PollAnswersCon {
  edges: [PollAnswersConEdge]!
  pageInfo: PageInfo!
}

type PollAnswersConEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersConnection {
  edges: [PollAnswersConnectionEdge]!
  pageInfo: PageInfo!
}

type PollAnswersConnectionEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersUsingOffsetConnection {
  edges: [PollAnswersUsingOffsetConnectionEdge]!
  pageInfo: PageInfo!
}

type PollAnswersUsingOffsetConnectionEdge {
  cursor: String!
  node: Answer!
}

type Query {
  batchNumbers(after: ID, before: ID, first: Int, last: Int): QueryBatchNumbersConnection!
  extraNode: Node
  inputGlobalID(id: ID!, inputObj: GlobalIDInput!, normalId: ID!): String!
  moreNodes: [Node]!
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
  numbers(after: ID, before: ID, first: Int, last: Int): QueryNumbersConnection!
  poll(id: Int!): Poll
  pollIds: [ID!]!
  polls: [Poll!]!
  pollsConnection(after: ID, before: ID, first: Int, last: Int): QueryPollsConnection!
  sharedConnection(after: ID, before: ID, first: Int, last: Int): SharedConnection
}

type QueryBatchNumbersConnection {
  connectionField: Int
  edges: [QueryBatchNumbersConnectionEdge]!
  pageInfo: PageInfo!
}

type QueryBatchNumbersConnectionEdge {
  cursor: String!
  edgeField: Int
  node: BatchNumber!
}

type QueryNumbersConnection {
  edges: [QueryNumbersConnectionEdge]!
  pageInfo: PageInfo!
}

type QueryNumbersConnectionEdge {
  cursor: String!
  node: Number!
}

type QueryPollsConnection {
  edges: [QueryPollsConnectionEdge]!
  extraConnectionField: String!
  pageInfo: PageInfo!
}

type QueryPollsConnectionEdge {
  cursor: String!
  extraEdgeField: Int!
  node: Poll!
}

type SharedConnection {
  edges: [SharedEdge]!
  pageInfo: PageInfo!
}

type SharedEdge {
  cursor: String!
  node: Number!
}
"
`;

exports[`relay example schema generates expected schema with globalConnectionFields 1`] = `
"type Answer {
  count: Int!
  id: ID!
  value: String!
}

type BatchNumber implements Node {
  id: ID!
  number: Int!
}

\\"\\"\\"input type\\"\\"\\"
input CustomInputName {
  id: ID!
}

\\"\\"\\"output type\\"\\"\\"
type CustomOutputName {
  itWorked: Boolean!
}

input ExampleMutationInput {
  id: ID!
}

type ExampleMutationPayload {
  itWorked: Boolean!
}

input GlobalIDInput {
  circular: GlobalIDInput
  circularWithoutGlobalIds: NoGlobalIDInput
  id: ID!
  idList: [ID!]!
}

type Mutation {
  answerPoll(answer: Int!, id: ID!): Poll!
  createPoll(answers: [String!]!, question: String!): Poll!
  exampleMutation(
    \\"\\"\\"mutation input arg\\"\\"\\"
    input: ExampleMutationInput!
  ): ExampleMutationPayload!

  \\"\\"\\"mutation field\\"\\"\\"
  exampleWithDescriptions(
    \\"\\"\\"mutation input arg\\"\\"\\"
    customInput: CustomInputName!
  ): CustomOutputName!
}

input NoGlobalIDInput {
  circular: NoGlobalIDInput
  id: ID
}

\\"\\"\\"node type\\"\\"\\"
interface Node {
  id: ID!
}

type Number implements Node {
  id: ID!
  number: Int!
}

\\"\\"\\"page info type\\"\\"\\"
type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

type Poll implements Node {
  answers: [Answer!]!
  answersConnection(after: String, before: String, first: Int, last: Int): PollAnswersConnection!
  answersUsingOffset(after: String, before: String, first: Int, last: Int): PollAnswersUsingOffsetConnection!
  answersWithoutHelpers(after: String, before: String, first: Int, last: Int): PollAnswersCon!
  id: ID!
  question: String!
  updatedAt: String!
}

type PollAnswersCon {
  edges: [PollAnswersConEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type PollAnswersConEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersConnection {
  edges: [PollAnswersConnectionEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type PollAnswersConnectionEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersUsingOffsetConnection {
  edges: [PollAnswersUsingOffsetConnectionEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type PollAnswersUsingOffsetConnectionEdge {
  cursor: String!
  node: Answer!
}

type Query {
  batchNumbers(after: String, before: String, first: Int, last: Int): QueryBatchNumbersConnection!
  extraNode: Node
  inputGlobalID(id: ID!, inputObj: GlobalIDInput!, normalId: ID!): String!
  moreNodes: [Node]!

  \\"\\"\\"node query\\"\\"\\"
  node(id: ID!): Node

  \\"\\"\\"nodes query\\"\\"\\"
  nodes(ids: [ID!]!): [Node]!
  numbers(after: String, before: String, first: Int, last: Int): QueryNumbersConnection!
  poll(id: Int!): Poll
  pollIds: [ID!]!
  polls: [Poll!]!
  pollsConnection(after: String, before: String, first: Int, last: Int): QueryPollsConnection!
  sharedConnection(after: String, before: String, first: Int, last: Int): SharedConnection
}

type QueryBatchNumbersConnection {
  edges: [QueryBatchNumbersConnectionEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type QueryBatchNumbersConnectionEdge {
  cursor: String!
  node: BatchNumber!
}

type QueryNumbersConnection {
  edges: [QueryNumbersConnectionEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type QueryNumbersConnectionEdge {
  cursor: String!
  node: Number!
}

type QueryPollsConnection {
  edges: [QueryPollsConnectionEdge]!
  extraConnectionField: String!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type QueryPollsConnectionEdge {
  cursor: String!
  extraEdgeField: Int!
  node: Poll!
}

type SharedConnection {
  edges: [SharedEdge]!
  pageInfo: PageInfo!
  totalCount: Int!
  totalCount2: Int!
}

type SharedEdge {
  cursor: String!
  node: Number!
}
"
`;

exports[`relay example schema queries query array based connections 1`] = `
Object {
  "data": Object {
    "batchNumbers": Object {
      "connectionField": 3,
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "edgeField": 0,
          "node": Object {
            "number": 0,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "edgeField": 1,
          "node": Object {
            "number": 1,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "edgeField": 2,
          "node": Object {
            "number": 2,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "beforeAndAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
      },
    },
    "firstAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
      },
    },
    "firstAfterAtEnd": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
          "node": Object {
            "number": 198,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
          "node": Object {
            "number": 199,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
      },
    },
    "firstAndLast": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
      },
    },
    "lastAtBegining": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "lastBefore": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": Object {
            "number": 7,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
      },
    },
  },
}
`;

exports[`relay example schema queries query node and nodes on query type 1`] = `
Object {
  "data": Object {
    "batchNode": Object {
      "number": 0,
    },
    "extraNode": Object {
      "__typename": "Number",
      "number": 2,
    },
    "moreNodes": Array [
      Object {
        "__typename": "Number",
        "number": 2,
      },
      Object {
        "__typename": "Number",
        "number": 10,
      },
    ],
    "node": Object {
      "id": "TnVtYmVyOjI=",
      "number": 2,
    },
    "node2": Object {
      "number": 2,
    },
    "nodes": Array [
      Object {
        "number": 4,
      },
      Object {
        "number": 2,
      },
      Object {
        "number": 2,
      },
      Object {
        "batchNumber": 0,
      },
    ],
  },
}
`;

exports[`relay example schema queries query offset based connections 1`] = `
Object {
  "data": Object {
    "beforeAndAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
      },
    },
    "firstAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
      },
    },
    "firstAfterAtEnd": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
          "node": Object {
            "number": 198,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
          "node": Object {
            "number": 199,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
      },
    },
    "firstAndLast": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
      },
    },
    "lastAtBegining": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "lastBefore": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": Object {
            "number": 7,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
      },
    },
    "numbers": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "node": Object {
            "number": 1,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "sharedConnection": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "id": "TnVtYmVyOjA=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "node": Object {
            "id": "TnVtYmVyOjE=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "id": "TnVtYmVyOjI=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "id": "TnVtYmVyOjM=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "id": "TnVtYmVyOjQ=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo1",
          "node": Object {
            "id": "TnVtYmVyOjU=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo2",
          "node": Object {
            "id": "TnVtYmVyOjY=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": Object {
            "id": "TnVtYmVyOjc=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "id": "TnVtYmVyOjg=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "id": "TnVtYmVyOjk=",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMA==",
          "node": Object {
            "id": "TnVtYmVyOjEw",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMQ==",
          "node": Object {
            "id": "TnVtYmVyOjEx",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMg==",
          "node": Object {
            "id": "TnVtYmVyOjEy",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMw==",
          "node": Object {
            "id": "TnVtYmVyOjEz",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxNA==",
          "node": Object {
            "id": "TnVtYmVyOjE0",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxNQ==",
          "node": Object {
            "id": "TnVtYmVyOjE1",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxNg==",
          "node": Object {
            "id": "TnVtYmVyOjE2",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxNw==",
          "node": Object {
            "id": "TnVtYmVyOjE3",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOA==",
          "node": Object {
            "id": "TnVtYmVyOjE4",
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOQ==",
          "node": Object {
            "id": "TnVtYmVyOjE5",
          },
        },
      ],
    },
  },
}
`;
